docs/proposals/deployment.md

```diff
@@ -0,0 +1,269 @@
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The maximum number of pods that can be unavailable during the update.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
+  // Absolute number is calculated from percentage by rounding up.
+  // This can not be 0 if MaxSurge is 0.
+  // By default, a fixed value of 1 is used.
+  // Example: when this is set to 30%, the old RC can be scaled down by 30%
+  // immediately when the rolling update starts. Once new pods are ready, old RC
+  // can be scaled down further, followed by scaling up the new RC, ensuring
+  // that at least 70% of original number of pods are available at all times
+  // during the update.
+  MaxUnavailable IntOrString
+
+  // The maximum number of pods that can be scheduled above the original number of
+  // pods.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at
+  // the start of the update (ex: 10%). This can not be 0 if MaxUnavailable is 0.
+  // Absolute number is calculated from percentage by rounding up.
+  // By default, a value of 1 is used.
+  // Example: when this is set to 30%, the new RC can be scaled up by 30%
+  // immediately when the rolling update starts. Once old pods have been killed,
+  // new RC can be scaled up further, ensuring that total number of pods running
+  // at any time during the update is atmost 130% of original pods.
+  MaxSurge IntOrString
+
+  // Minimum number of seconds for which a newly created pod should be ready
+  // without any of its container crashing, for it to be considered available.
+  // Defaults to 0 (pod will be considered available as soon as it is ready)
+  MinReadySeconds int
+}
+
+type DeploymentStatus struct {
+  // Total number of ready pods targeted by this deployment (this
+  // includes both the old and new pods).
+  Replicas int
+
+  // Total number of new ready pods with the desired template spec.
+  UpdatedReplicas int
+}
+
+```
+
+### Controller
+
+#### Deployment Controller
+
+The DeploymentController will make Deployments happen.
+It will watch Deployment objects in etcd.
+For each pending deployment, it will:
+
+1. Find all RCs whose label selector is a superset of DeploymentSpec.Selector.
+   - For now, we will do this in the client - list all RCs and then filter the
+     ones we want. Eventually, we want to expose this in the API.
+2. The new RC can have the same selector as the old RC and hence we add a unique
+   selector to all these RCs (and the corresponding label to their pods) to ensure
+   that they do not select the newly created pods (or old pods get selected by
+   new RC).
+   - The label key will be "deployment.kubernetes.io/podTemplateHash".
+   - The label value will be hash of the podTemplateSpec for that RC without
+     this label. This value will be unique for all RCs, since PodTemplateSpec should be unique.
```

Comment:
I think we use Adler32 to do the limited hash today.

On Sep 3, 2015, at 9:05 AM, Kamal Marhubi <notifications@github.com> wrote:

In docs/proposals/deployment.md
<https://github.com/kubernetes/kubernetes/pull/12236#discussion_r38642754>:

> +#### Deployment Controller
> +
> +The DeploymentController will make Deployments happen.
> +It will watch Deployment objects in etcd.
> +For each pending deployment, it will:
> +
> +1. Find all RCs whose label selector is a superset of DeploymentSpec.Selector.
> +   - For now, we will do this in the client - list all RCs and then filter the
> +     ones we want. Eventually, we want to expose this in the API.
> +2. The new RC can have the same selector as the old RC and hence we add a unique
> +   selector to all these RCs (and the corresponding label to their pods) to ensure
> +   that they do not select the newly created pods (or old pods get selected by
> +   new RC).
> +   - The label key will be "deployment.kubernetes.io/podTemplateHash".
> +   - The label value will be hash of the podTemplateSpec for that RC without
> +     this label. This value will be unique for all RCs, since PodTemplateSpec should be unique.

Will this be hashing the textual representation (JSON?) or the
PodTemplateSpec object? In the former case, I suggest using something like
tent/canonical-json <https://github.com/tent/canonical-json-go> which is
used by the TUF <http://theupdateframework.com/> implementation in
docker/notary <https://github.com/docker/notary>.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12236/files#r38642754>.